pre-sync/oci-image-verification/main.go

```diff
@@ -0,0 +1,145 @@
+
+// Function to extract annotations from JSON
+func getAnnotations(raw []byte) (map[string]string, error) {
+	var metadata map[string]interface{}
+	if err := json.Unmarshal(raw, &metadata); err != nil {
+		return nil, err
+	}
+	if annotations, ok := metadata["metadata"].(map[string]interface{})["annotations"].(map[string]interface{}); ok {
+		annotationsMap := make(map[string]string)
+		for k, v := range annotations {
+			annotationsMap[k] = fmt.Sprintf("%v", v)
+		}
+		return annotationsMap, nil
+	}
+	return nil, fmt.Errorf("no annotations found")
+}
+
+func validateImage(image string) error {
+	cmd := exec.Command("cosign", "verify", image, "--key", "/cosign-key/cosign.pub")
+	log.Printf("command %s", cmd.String())
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cosign verification failed: %s, output: %s", err, string(output))
+	}
+	return nil
+}
+
+func auth() error {
+	authMutex.Lock()         // Acquire the lock before accessing shared resources
+	defer authMutex.Unlock() // Release the lock when the function exits
+
+	if authorized {
+		log.Printf("skip authorizing docker")
+		return nil
+	}
+	gcloudCmd := exec.Command("gcloud", "auth", "print-access-token")
+	accessTokenBytes, err := gcloudCmd.Output()
+	if err != nil {
+		return fmt.Errorf("Error fetching access token: %v\n", err)
+	}
+	// Convert the output to a string and trim any trailing newline
+	accessToken := string(accessTokenBytes)
+	accessToken = accessToken[:len(accessToken)-1] // Remove the trailing newline
+
+	cmd := exec.Command("cosign", "login", "us-central1-docker.pkg.dev", "-u", "oauth2accesstoken", "-p", accessToken)
+	log.Printf("command %s", cmd.String())
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("gcloud auth ar failed: %s, output: %s", err, string(output))
+	}
+	log.Printf("result: %s, authorization done", string(output))
+	authorized = true
+	return nil
+}
+
+func handleWebhook(w http.ResponseWriter, r *http.Request) {
+
+	// Read the body
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Failed to read request body: %v", err)
+		http.Error(w, "Failed to read request body", http.StatusBadRequest)
+		return
+	}
+
+	var admissionReview admissionv1.AdmissionReview
+	if err := json.Unmarshal(body, &admissionReview); err != nil {
+		log.Printf("Failed to unmarshal admission review: %v", err)
+		http.Error(w, "Failed to unmarshal admission review", http.StatusBadRequest)
+		return
+	}
+
+	// Extract old and new annotations
+	oldAnnotations, err := getAnnotations(admissionReview.Request.OldObject.Raw)
+	if err != nil {
+		log.Printf("Failed to extract old annotations: %v", err)
+	}
+
+	newAnnotations, err := getAnnotations(admissionReview.Request.Object.Raw)
+	if err != nil {
+		log.Printf("Failed to extract new annotations: %v", err)
+	}
+
+	// Log old and new annotations for comparison
+	log.Printf("Old Annotations: %v", oldAnnotations)
```

Comment:
Suggest using klog and make it a lower level log.

Reply:
Done